src/orion/algo/asha.py

```diff
@@ -0,0 +1,297 @@
+        Default: 1
+    reduction_factor: int
+        The factor by which ASHA promotes trials. If the reduction factor is 4,
+        it means the number of trials from one fidelity level to the next one is roughly
+        divided by 4, and each fidelity level has 4 times more resources than the prior one.
+        Default: 4
+    num_brackets: int
+        Using a grace period that is too small may bias ASHA too strongly towards
+        fast converging trials that do not lead to best results at convergence (stagglers). To
+        overcome this, you can increase the number of brackets, which increases the amount of
+        resource required for optimisation but decreases the bias towards stragglers.
+        Default: 1
+
+    """
+
+    def __init__(self, space, seed=None, max_resources=100, grace_period=1, reduction_factor=4,
+                 num_brackets=1):
+        super(ASHA, self).__init__(
+            space, seed=seed, max_resources=max_resources, grace_period=grace_period,
+            reduction_factor=reduction_factor, num_brackets=num_brackets)
+
+        if reduction_factor < 2:
+            raise AttributeError("Reduction factor for ASHA needs to be at least 2.")
+
+        self.trial_info = {}  # Stores Trial -> Bracket
+
+        # Tracks state for new trial add
+        self.brackets = [
+            Bracket(self, grace_period, max_resources, reduction_factor, s)
+            for s in range(num_brackets)
+        ]
+
+    def seed_rng(self, seed):
+        """Seed the state of the random number generator.
+
+        :param seed: Integer seed for the random number generator.
+        """
+        self.rng = numpy.random.RandomState(seed)
+
+    @property
+    def state_dict(self):
+        """Return a state dict that can be used to reset the state of the algorithm."""
+        return {'rng_state': self.rng.get_state()}
+
+    def set_state(self, state_dict):
+        """Reset the state of the algorithm based on the given state_dict
+
+        :param state_dict: Dictionary representing state of an algorithm
+        """
+        self.seed_rng(0)
+        self.rng.set_state(state_dict['rng_state'])
+
+    def suggest(self, num=1):
+        """Suggest a `num` of new sets of parameters.
+
+        Promote a trial if possible, otherwise randomly draw samples from the space and
+        randomly assign to a bracket.
+
+        :param num: how many sets to be suggested.
+
+        .. note:: New parameters must be compliant with the problem's domain
+           `orion.algo.space.Space`.
+        """
+        if num > 1:
+            raise ValueError("ASHA should suggest only one point.")
+
+        for bracket in self.brackets:
+            candidate = bracket.update_rungs()
+
+            if candidate:
+                logger.debug('Promoting')
+                return [candidate]
+
+        for _attempt in range(100):
+            point = list(self.space.sample(1, seed=tuple(self.rng.randint(0, 1000000, size=3)))[0])
```

Comment:
Because we are using scipy distributions for priors in `Space`, we can't create a random number generator with a seed, we must pass a seed value when we sample. This is why we need to pass a different number each time, so to do this we sample the seeds from the algo's RNG object. It should be reproducible because all distributed algos will always go through the same sequence of seeds. In practice it won't be fully reproducible however because the order of trial completion will vary based on the computing system, but that is another issue.